keps/sig-network/4559-kubelet-probes/README.md

```diff
@@ -0,0 +1,1037 @@
+  to travel over the network from kubelet to the pod. This, of course,
+  only works if there is nothing in the pod network blocking kubelet's
+  access to those pods. When we added NetworkPolicy to Kubernetes, we
+  didn't want users to have to manually add rules allowing access from
+  Kubelet (and didn't provide any way for them to write such a rule
+  anyway), so we just said that kubelet probes are not subject to
+  NetworkPolicy, and the NetworkPolicy implementation needs to make
+  that work. In almost all cases, this is implemented by allowing
+  _all_ node-level processes and host-network pods on the node to
+  reach _all_ ports on _all_ pod-network pods on the node, which users
+  probably do not want (and may be unaware of). SIG Network would like
+  to avoid continuing to have this problem in [AdminNetworkPolicy] and
+  any [future successor to NetworkPolicy].
+
+- **Contraints on Network Architecture Imposed by Pod Probes**: The
+  fact that kubelet must be able to reach every pod on its node
+  implies that every pod on a given node must have a distinct IP
+  address. While that is certainly required anyway if every pod in the
+  cluster is expected to be able to be able to talk to every other
+  pod, there are clusters where that is actually an "anti-requirement"
+  (e.g., [multi-tenant scenarios]), and in some of these cases, people
+  have wanted to reuse the same pod IPs for different pods. Although
+  it would be possible to deploy a Kubernetes cluster with a CNI
+  plugin, service proxy and NetworkPolicy implementation that properly
+  handle non-unique pod IPs, there is currently no way to make
+  network-based pod probes work correctly in such an environment.
+
+  The proposed [Multi-Network feature] raises similar issues, in that
+  in addition to the possibility of networks with overlapping IPs, it
+  also potentially involves networks that cannot be reached from the
+  host-network namespace.
+
+- **Dual-Stack Pods**: Probes are currently defined to use the pod's
+  primary IP family, but there are cases where you might want to do an
+  IPv6 probe in an IPv4-primary dual-stack cluster. ([kubernetes
+  #101324])
+
+- **Unclear Semantics for Network Probes**: It was never defined
+  whether probes were intended to check "is the server process in the
+  pod running?" or "is the pod reachable over the network?" or both.
+  While many people would like them to mean the latter, this has
+  turned out to not really work in many cases. (In particular, the
+  fact that kubelet probes are exempt from NetworkPolicy means that a
+  probe may succeed even when no other pods would be able to reach the
+  pod because [the NetworkPolicy has not yet been fully programmed].)
+  This led to the creation of [readiness gates] to provide better pod
+  readiness information.
+
+- **Exec Probe Performance (?)**: [Exec probes are much, much slower than
+  network probes]. Apparently this slowness is a somewhat inevitable
```

Comment:
I was unaware of this slowness problem.
Is there more context for me to go read up on it? An issue perhaps?

Reply:
Oh, I see below that "Figure out details of exec probe performance problem" is listed

Reply:
The brackets in the text here indicate that that's a link, with the actual URL done "footnote-style" a bit further down in the text. In this case, the link is to a blog post, "[Exec Probes: A Story about Experiment and Relevant Problems](https://medium.com/netcracker/exec-probes-a-story-about-experiment-and-relevant-problems-12de616c0a76)". I don't have much more context than that, other than my guesses in this paragraph. (Though also, I don't think it was ever confirmed that this is a problem with exec probes in all CRI runtimes, or if it's basically just a containerd bug.)

Reply:
I can add a note to that effect that we have not had any specific issues opened in the past around this but listing the problem here based on the post.

Reply:
we need someone from Node to evaluate this or we have the risk of operating with biased information, @samuelkarp can you help us to understand better this performance problem with Exec probes? is this a general and unsolvable problem?

Reply:
exec with a new task is not as expensive as running a container..but still, it's not an inexpensive operation either.. and the usual problem is there are a plethora of exec probes that are not cycle optimized because of the default timeouts and repeat duration values.. And when the system gets stressed the probe failures just make it worse.

We could probably figure out more optimizations... sure that's worthy of trying even if just for non-probe exec. Would be interesting to see if there is a difference between crio/containerd or any other improvements since the last time we checked. 

Though, I held out hope for a different probe api that did not lean on/overload exec and the streaming api.